src/AMAMM.sol

```diff
@@ -2,69 +2,176 @@
+        }
+
+        if (_getDeposit(id, _epoch) < deposit) {
+            _userBalance[poolEpochManager[id][_epoch].bidder] += _getDeposit(id, _epoch); //Increase user balance of losing bidder
+            poolEpochManager[id][_epoch] = Bid({bidder: msgSender, payload: payload, rent: rent});
+        } else {
+            _userBalance[msgSender] += deposit; //Increase balance if bidder is not top bidder
+        }
+
+        _updateEpochBids(id);
     }
 
-    function afterSwap(
-        address /* sender **/,
-        PoolKey calldata key,
-        IPoolManager.SwapParams calldata params,
-        BalanceDelta delta,
-        bytes calldata /* hookData **/
-    ) external override poolManagerOnly returns (bytes4, int128) {
-        // fee will be in the unspecified token of the swap
-        bool specifiedTokenIs0 = (params.amountSpecified < 0 ==
-            params.zeroForOne);
-        (Currency feeCurrency, int128 swapAmount) = (specifiedTokenIs0)
-            ? (key.currency1, delta.amount1())
-            : (key.currency0, delta.amount0());
-        // if fee is on output, get the absolute output amount
-        if (swapAmount < 0) swapAmount = -swapAmount;
-
-        uint256 feeAmount = (uint128(swapAmount) * SWAP_FEE_BIPS) / TOTAL_BIPS;
-        // TODO: change address(this) to the ePoche manager address
-        poolManager.take(feeCurrency, address(this), feeAmount);
-
-        return (IHooks.afterSwap.selector, feeAmount.toInt128());
+    /// @inheritdoc IAmAmm
+    function withdrawBalance(PoolId id, uint128 _amount) external virtual override returns (uint128) {
+        /// -----------------------------------------------------------------------
+        /// Validation
+        /// -----------------------------------------------------------------------
+
+        address msgSender = LibMulticaller.senderOrSigner();
+
+        if (!_amAmmEnabled(id) || _userBalance[msgSender] < _amount) {
+            revert AmAmm__InvalidBid();
+        }
+
+        /// -----------------------------------------------------------------------
+        /// State updates
+        /// -----------------------------------------------------------------------
+
+        unchecked {
+            _userBalance[msgSender] -= _amount;
+
+            _updateEpochBids(id);
+        }
+
+        return _amount;
     }
+
+    /// -----------------------------------------------------------------------
+    /// Internal helpers
+    /// -----------------------------------------------------------------------
+
+    /// @dev Charges rent
+    function _updateEpochBids(PoolId id) internal virtual {
+        uint40 currentEpoch = _getEpoch(id, block.timestamp);
+
+        // Early return if the pool has already been updated in this epoch
+        if (_lastUpdatedEpoch[id] == currentEpoch) {
+            return;
+        }
+
+        // Loop through each epoch that has passed
+        for (uint40 epoch = _lastUpdatedEpoch[id] + 1; epoch <= currentEpoch; epoch++) {
```

Comment:
For loop is inefficient and one more issue here is it also needs to check if there's an existing bidder in each epoch and if rent is more than the epoch.rent, it also need to update the bidder. 

However I would like to propose a simpler implementation that don't need the for loop. As the user is paying for k epochs. So essentially that user is the manager for k epochs from lastUpdatedEpoch. Therefore, there would be just two scenario that if a new bid is <= lastUpdatedEpoch + k or more than. So when less than we could compare the rent and update bidder and new rent. If more than we could just put in the new bidder.